tendermint/src/tendermint.rs

```diff
@@ -390,52 +391,104 @@ impl<TProtocol: Protocol> Tendermint<TProtocol> {
+                                self.future_contributions.retain(|&round, _contributors| {
+                                    round > self.state.current_round
+                                });
+                                self.future_round_messages
+                                    .retain(|_, message| message.tag.0 > self.state.current_round);
+
+                                // Return to account the state change
+                                return None;
+                            }
+                        }
+                    }
                 }
-            } else if let Some(sender) = self.aggregation_senders.get(&id) {
-                match sender.try_send(message.aggregation) {
-                    Ok(()) => {}
-                    Err(mpsc::error::TrySendError::Closed(_aggregation_message)) => {
-                        panic!("Aggregations must not be dropped.");
+            }
+
+            // If no future round aggregation verification is currently running, start a new one.
+            if self.future_round_verification.is_none() && !self.future_round_messages.is_empty() {
+                // TODO: choose at random
+                let (_, message) = self.future_round_messages.pop_first().unwrap();
+                let (round, step) = message.tag;
+                let contributors = message.aggregation.all_contributors();
+                let verification = self
+                    .protocol
+                    .verify_aggregation_message(round, step, message.aggregation)
+                    .map(move |result| result.map(move |()| (round, contributors)));
+                self.future_round_verification = Some(Box::pin(verification));
+            }
+
+            // Finally, check if we have new messages.
+            //
+            // the state changes if there is at least one message to process as they are kept for future reference.
+            if let Poll::Ready(Some(mut message)) = self.level_update_stream.poll_next_unpin(cx) {
+                did_something = true;
+                let id = message.tag;
+                if id.0 > self.state.current_round {
+                    // Queue the message for future round aggregation verification.
+                    match self
+                        .future_round_messages
+                        .entry(message.aggregation.sender())
+                    {
+                        Entry::Vacant(v) => {
+                            v.insert(message);
+                        }
+                        Entry::Occupied(mut o) => {
+                            // Generally: Newer is better. If the new message comes
+                            // from the same round as a previous one, assume that
+                            // it is better.
+                            if id.0 >= o.get().tag.0 {
+                                o.insert(message);
+                            }
+                        }
                     }
```

Comment:
I like the Entry API better here as well.
```suggestion
                    match self
                        .future_round_messages
                        .entry(message.aggregation.sender())
                        .and_modify(|m| {
                            if id.0 > m.tag.0 {
                                *m = message
                            }
                        })
                        .or_insert(message)
```

Reply:
I also like that better, but it's not possible due to ownership rules, unfortunately. The closure would take ownership of `message`, but it's needed afterwards.